### PR TITLE
Analtics pipeline

### PR DESCRIPTION
**Pull request by dariorussi**
## Description 

Initial implementation of the analytics pipeline. 
I am thinking it would be best to start reviewing this and committing sooner as opposed to get this in a great shape and put it up for review only then. It's getting bigger and there is still quite a bit to do and it will only get bigger and bigger, so we better start moving.
The overall idea is here though. So please take a look and let me know if you see issues that you want solved or done differently, or just broken.
@sadhansood, @bschwab and I talked about this and trying to get it in seems important, so others can work on it as well, and we can iterate.
Nothing uses it for now. Last I tried it, on my machine, with a local fullnode, it collect checkpoints and runs. So it's usable.
Next step is to write a csv writer and start the task that reads those files. Sadhan is implementing the uploader to snowflake.

Schema are likely to change but that is ok.

## Test Plan 

Nothing for now and we need to come up with something, not sure what though. Suggestion welcome.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-analytics-indexer/src/tables.rs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    pub(crate) transaction_kind: String,
+    pub(crate) transaction_count: u64,
+    pub(crate) execution_success: bool,
+    // object info
+    pub(crate) input: u64,
+    pub(crate) shared_input: u64,
+    pub(crate) gas_coins: u64,
+    // objects are broken up in created, mutated and deleted.
+    // No wrap or unwrap information is provided
+    pub(crate) created: u64,
```
Comment (gegaowp) [inline crates/sui-analytics-indexer/src/tables.rs:58]: `u64` means count of created objects?

File: crates/sui-analytics-indexer/src/tables.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    pub(crate) mutated: u64,
+    pub(crate) deleted: u64,
+    // PTB info
+    pub(crate) move_calls: u64,
+    pub(crate) packages: BTreeSet<String>,
```
Comment (gegaowp) [inline crates/sui-analytics-indexer/src/tables.rs:63]: what does `packages` of a txn mean?

File: crates/sui-analytics-indexer/src/tables.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    // indexes
+    pub(crate) object_id: String,
+    pub(crate) version: Option<u64>,
+    pub(crate) transaction_digest: String,
+    pub(crate) checkpoint: u64,
+    pub(crate) epoch: u64,
+    pub(crate) timestamp_ms: u64,
+    // input/output information
+    pub(crate) input_kind: Option<InputObjectKind>,
+    pub(crate) object_status: Option<ObjectStatus>,
```
Comment (gegaowp) [inline crates/sui-analytics-indexer/src/tables.rs:166]: what about wrapped and unwrapped?

File: crates/sui-analytics-indexer/src/tables.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    pub(crate) bcs: Vec<u8>,
+}
+
+// Objects used and manipulated in a transaction.
+// Both input object and objects in effects are reported here with the proper
+// input kind (for input objects) and status (for objets in effects).
+// An object may appear twice as an input and output object. In that case, the
+// version will be different.
+#[derive(Serialize)]
+pub(crate) struct TransactionObjectEntry {
```
Comment (gegaowp) [inline crates/sui-analytics-indexer/src/tables.rs:156]: curious what's the current use-case of this table?
(on the OLTP end, it's mainly used as indexing table for queryTxnBlocks)

File: crates/sui-analytics-indexer/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+anyhow.workspace = true
+async-trait.workspace = true
+axum.workspace = true
+backoff.workspace = true
+bcs.workspace = true
+chrono.workspace = true
+clap.workspace = true
+csv.workspace = true
+tap.workspace = true
+diesel.workspace = true
```
Comment (gegaowp) [inline crates/sui-analytics-indexer/Cargo.toml:20]: nit: diesel seems not used?

**Review by gegaowp — COMMENTED** [review]
Looks good, thanks for kicking it off! I have a couple of high-level verification questions to make sure I follow:
- looks like this analytical indexer will fetch data and write to csv files, how are these files read by Snowflake, is it automated?
- presumably we will need a separate binary / pod besides the indexer writer pod to run this?

File: crates/sui-analytics-indexer/src/tables.rs (hunk 2 above)
Comment (dariorussi) [reply crates/sui-analytics-indexer/src/tables.rs:63]: in a sense I am not sure yet (lol), it is possibly a list of packages (comma separated or similar) used in the transaction.
The idea is to have an easy way via SQL (with regex match) to find all transactions that work against a certain package.
Thoughts?

File: crates/sui-analytics-indexer/src/tables.rs (hunk 3 above)
Comment (dariorussi) [reply crates/sui-analytics-indexer/src/tables.rs:166]: planning to talk to more people about this and glad you asked.
I could not find any reason to have/use the wrapped/unwrapped/etc. info and they are folded into created/mutated/deleted. 
What do you think?

File: crates/sui-analytics-indexer/Cargo.toml (hunk 5 above)
Comment (dariorussi) [reply crates/sui-analytics-indexer/Cargo.toml:20]: yeah thanks, will remove and probably more. I was looking for a tool that could tell unused dependencies but could not find any. If you know something let me know please

**Comment by dariorussi** [comment]
thanks @gegaowp, as for the general questions:

> looks like this analytical indexer will fetch data and write to csv files, how are these files read by Snowflake, is it automated?

The csv files is more an example that how we will do things. Though it is an option. The plan is to have more "writer" that can save to different formats, though that story is to be built. The snowflake upload is a work in progress and will appear here soon and be part of this binary.

> presumably we will need a separate binary / pod besides the indexer writer pod to run this?

correct, that is the current plan of record. We have a conversation open with PE to sort that out. That seems to right approach to me. Do you have thoughts about it?
Certainly curious and eager to hear your thoughts

**Review by sadhansood — APPROVED** [review]

**Pull request merged**